sdk/src/main/java/com/hedera/hashgraph/sdk/TopicMessageQuery.java

```diff
@@ -1,110 +1,199 @@
-import com.hedera.hashgraph.sdk.proto.mirror.ConsensusTopicQuery;
-import com.hedera.hashgraph.sdk.proto.mirror.ConsensusTopicResponse;
+import com.hedera.hashgraph.sdk.proto.Timestamp;
 import io.grpc.CallOptions;
 import io.grpc.ClientCall;
 import io.grpc.Status;
 import io.grpc.StatusRuntimeException;
 import io.grpc.stub.ClientCalls;
 import io.grpc.stub.StreamObserver;
-import java8.util.function.BiConsumer;
-import java8.util.function.Consumer;
-import org.threeten.bp.Instant;
-
-import javax.annotation.Nullable;
+import java.time.Duration;
 import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.Objects;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
+import java.util.regex.Pattern;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.threeten.bp.Instant;
+
+import com.hedera.hashgraph.sdk.proto.TransactionID;
+import com.hedera.hashgraph.sdk.proto.mirror.ConsensusServiceGrpc;
+import com.hedera.hashgraph.sdk.proto.mirror.ConsensusTopicQuery;
+import com.hedera.hashgraph.sdk.proto.mirror.ConsensusTopicResponse;
 
 public final class TopicMessageQuery {
-    private final ConsensusTopicQuery.Builder builder;
 
-    @Nullable
-    private BiConsumer<Throwable, TopicMessage> errorHandler;
+    private static final Logger LOGGER = LoggerFactory.getLogger(TopicMessageQuery.class);
+    private static final Pattern RST_STREAM = Pattern
+            .compile(".*(rst.stream.*internal.error|internal.error.*rst.stream).*",
+                    Pattern.CASE_INSENSITIVE | Pattern.DOTALL);
+
+    private final ConsensusTopicQuery.Builder builder;
+    private Runnable completionHandler = this::onComplete;
+    private BiConsumer<Throwable, TopicMessage> errorHandler = this::onError;
+    private int maxAttempts = 10;
+    private Duration maxBackoff = Duration.ofSeconds(8L);
+    private Predicate<Throwable> retryHandler = this::shouldRetry;
 
     public TopicMessageQuery() {
         builder = ConsensusTopicQuery.newBuilder();
     }
 
     public TopicMessageQuery setTopicId(TopicId topicId) {
+        Objects.requireNonNull(topicId, "topicId must not be null");
         builder.setTopicID(topicId.toProtobuf());
         return this;
     }
 
     public TopicMessageQuery setStartTime(Instant startTime) {
+        Objects.requireNonNull(startTime, "startTime must not be null");
         builder.setConsensusStartTime(InstantConverter.toProtobuf(startTime));
         return this;
     }
 
     public TopicMessageQuery setEndTime(Instant endTime) {
+        Objects.requireNonNull(endTime, "endTime must not be null");
         builder.setConsensusEndTime(InstantConverter.toProtobuf(endTime));
         return this;
     }
 
     public TopicMessageQuery setLimit(long limit) {
+        if (limit < 0) {
```

Comment:
We try to avoid doing parameter validation on the client side because the server requirements have often changed; I think it'd be best to simply allow the user to set the limit to any value, and let the server respond with an error instead.

Reply:
Removed